core/src/test/java/net/opentsdb/query/TestSemanticQuery.java

```diff
@@ -123,6 +124,137 @@ public void builder() throws Exception {
       fail("Expected IllegalArgumentException");
     } catch (IllegalArgumentException e) { }
   }
+
+  @Test
+  public void timestampPermutations() throws Exception {
+
+    NamedFilter filter = DefaultNamedFilter.newBuilder()
+        .setFilter(TagValueLiteralOrFilter.newBuilder()
+            .setFilter("web01")
+            .setTagKey("host")
+            .build())
+        .setId("f1")
+        .build();
+    List<QueryNodeConfig> graph = Lists.newArrayList(
+        DefaultTimeSeriesDataSourceConfig.newBuilder()
+            .setMetric(MetricLiteralFilter.newBuilder()
+                .setMetric("sys.cpu.user")
+                .build())
+            .setFilterId("f1")
+            .setId("m1")
+            .build());
+
+    SemanticQuery endTimeAfterCurrentTimeAndNoStartTime = SemanticQuery.newBuilder()
+        .setMode(QueryMode.SINGLE)
+        .setStart("")
+        .setEnd("9999999999")
+        .setTimeZone("America/Denver")
+        .addFilter(filter)
+        .setExecutionGraph(graph)
+        .addSerdesConfig(JsonV2QuerySerdesOptions.newBuilder()
+            .addFilter("ds")
+            .setId("serdes")
+            .build())
+        .build();
+
+    assertEquals(null, endTimeAfterCurrentTimeAndNoStartTime.getStart());
+    assertEquals(DateTime.currentTimeMillis() / 1000, endTimeAfterCurrentTimeAndNoStartTime.startTime().epoch());
+    assertEquals("9999999999", endTimeAfterCurrentTimeAndNoStartTime.getEnd());
+    assertEquals(9999999999L, endTimeAfterCurrentTimeAndNoStartTime.endTime().epoch());
+
+
+
+    SemanticQuery endTimeBeforeCurrentTimeAndNoStartTime = SemanticQuery.newBuilder()
+        .setMode(QueryMode.SINGLE)
+        .setStart("")
+        .setEnd("1000000000")
+        .setTimeZone("America/Denver")
+        .addFilter(filter)
+        .setExecutionGraph(graph)
+        .addSerdesConfig(JsonV2QuerySerdesOptions.newBuilder()
+            .addFilter("ds")
+            .setId("serdes")
+            .build())
+        .build();
+
+    assertEquals(null, endTimeBeforeCurrentTimeAndNoStartTime.getStart());
+    assertEquals(1000000000, endTimeBeforeCurrentTimeAndNoStartTime.startTime().epoch());
+    assertEquals("1000000000", endTimeBeforeCurrentTimeAndNoStartTime.getEnd());
+    assertEquals(DateTime.currentTimeMillis() / 1000, endTimeBeforeCurrentTimeAndNoStartTime.endTime().epoch());
```

Comment:
We'll need to mock these out or it can be indeterminate if the host is really really slow. As an example, core/src/test/java/net/opentsdb/query/execution/TestTimeBasedRoutingExecutor.java

Reply:
Are there any working examples?